File: src/main/javabak/multiblock/MultiblockBlockEntity.java
package redcrafter07.processed.multiblock;

import net.minecraft.client.multiplayer.ClientLevel;
import net.minecraft.core.BlockPos;
import net.minecraft.core.HolderLookup;
import net.minecraft.core.SectionPos;
import net.minecraft.core.particles.ParticleTypes;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.chunk.ChunkAccess;
import net.minecraft.world.level.chunk.status.ChunkStatus;
import net.neoforged.neoforge.items.IItemHandler;
import redcrafter07.processed.block.machine_abstractions.BlockSide;
import redcrafter07.processed.block.machine_abstractions.ProcessedMachine;
import redcrafter07.processed.network.MultiblockDestroyPacket;

import javax.annotation.Nullable;
import java.util.HashSet;
import java.util.Set;

import static redcrafter07.processed.ProcessedUtil.*;

public abstract class MultiblockBlockEntity extends ProcessedMachine {
    public MultiblockBlockEntity(BlockEntityType<?> type, BlockPos pos, BlockState blockState) {
        super(type, pos, blockState);
    }

    protected abstract MultiblockValidator validator();

    private int timeUntilNextCheck = 40;
    private boolean isAssembled = false;
    private @Nullable Set<BlockPos> blocks = null;
    private @Nullable Set<BlockPos> blocksOld = null;

    private void synchroniseWithCache() {
        if (blocksOld == null) return;
        Level level = getLevel();
        if (level == null) return;

        for (BlockPos block : blocksOld) {
            if (blocks != null && blocks.contains(block)) continue;
            ChunkAccess chunk = level.getChunk(SectionPos.blockToSectionCoord(block.getX()), SectionPos.blockToSectionCoord(block.getZ()), ChunkStatus.FULL, false);
            if (chunk == null) return;
            MultiblockCasingCache cache = MultiblockCasingCache.get(chunk);
            if (cache == null) continue;
            if (!cache.multiblockMap().containsKey(block)) continue;
            if (!cache.multiblockMap().get(block).equals(getBlockPos())) continue;
            cache.multiblockMap().remove(block);
            cache.set(chunk);
        }

        if (blocks == null) {
            blocksOld = null;
            return;
        }
        for (BlockPos block : blocks) {
            if (blocksOld.contains(block)) continue;
            ChunkAccess chunk = level.getChunk(SectionPos.blockToSectionCoord(block.getX()), SectionPos.blockToSectionCoord(block.getZ()), ChunkStatus.FULL, false);
            if (chunk == null) return;
            MultiblockCasingCache cache = MultiblockCasingCache.getOrDefault(chunk);
            cache.multiblockMap().put(block, getBlockPos());
            cache.set(chunk);
        }
        blocksOld = null;
    }

    @Override
    protected void loadAdditional(CompoundTag tag, HolderLookup.Provider registries) {
        super.loadAdditional(tag, registries);
        isAssembled = tag.getBoolean("isAssembled");
        if (tag.contains("blocks", 11)) {
            if (blocks == null) blocksOld = Set.of();
            else blocksOld = blocks;
            blocks = new HashSet<>();
            blocks.addAll(loadBlockPositions(tag.getIntArray("blocks")));
        } else {
            blocksOld = blocks == null ? Set.of() : blocks;
            blocks = null;
            isAssembled = false;
        }
    }

    @Override
    protected void saveAdditional(CompoundTag tag, HolderLookup.Provider registries) {
        super.saveAdditional(tag, registries);
        tag.putBoolean("isAssembled", isAssembled);
        if (blocks != null) tag.put("blocks", saveBlockPositions(blocks.stream().toList()));
    }

    public void onRemove(Level level) {
        if (blocks == null) return;
        if (level instanceof ServerLevel serverLevel && !level.isClientSide()) {
            MultiblockDestroyPacket packet = new MultiblockDestroyPacket(blocks.stream().toList(), getBlockPos());
            for (ServerPlayer player : serverLevel.players()) player.connection.send(packet);
        }
        for (BlockPos pos : blocks) {
            MultiblockCasingCache.removeCasing(level, pos);
            level.invalidateCapabilities(pos);
        }
    }

    public void tileTickCommon(Level level, BlockPos pos, BlockState state) {
    }

    public void tileTickServer(ServerLevel level, BlockPos pos, BlockState state) {
    }

    public void tileTickClient(ClientLevel level, BlockPos pos, BlockState state) {
    }

    @Override
    public final void serverTick(ServerLevel level, BlockPos pos, BlockState state) {
    }

    @Override
    public final void clientTick(ClientLevel level, BlockPos pos, BlockState state) {
    }

    @Override
    public final void commonTick(Level level, BlockPos pos, BlockState state) {
        if (!isAssembled || !isRunning()) return;
        tileTickCommon(level, pos, state);
        if (level instanceof ServerLevel serverLevel && !level.isClientSide()) tileTickServer(serverLevel, pos, state);
        else if (level instanceof ClientLevel clientLevel && level.isClientSide())
            tileTickClient(clientLevel, pos, state);
    }

    @Override
    public final void tickNoProcessing(Level level, BlockPos pos, BlockState state) {
        synchroniseWithCache();

        if (level instanceof ServerLevel && !level.isClientSide()) {
            if (isAssembled) return;
            if (timeUntilNextCheck > 0) {
                timeUntilNextCheck -= 1;
                return;
            }
            timeUntilNextCheck = 40;

            recheck();
        }
    }

    @Nullable
    @Override
    public final IItemHandler itemCapabilityForSide(@Nullable BlockSide side, BlockState state) {
        if (isAssembled) return super.itemCapabilityForSide(side, state);
        return null;
    }

    private @Nullable ServerLevel getLevelServer() {
        if (getLevel() instanceof ServerLevel sp && !sp.isClientSide) return sp;
        return null;
    }

    public void recheck() {
        ServerLevel serverLevel = getLevelServer();
        if (serverLevel == null) return;

        @Nullable Set<BlockPos> affectedBlocks = validator().getBlocks(serverLevel, getBlockPos(), getFacingDirection(getBlockState()));
        if (affectedBlocks != null) {
            for (BlockPos block : affectedBlocks) {
                @Nullable BlockPos controllerPos = MultiblockCasingCache.getControllerForCasing(serverLevel, block);
                if (controllerPos != null && controllerPos != getBlockPos()) {
                    affectedBlocks = null;
                    break;
                }
            }
        }

        if (blocks == null) blocks = Set.of();
        Set<BlockPos> old = blocks;
        blocks = null;
        var wasPreviouslyAssembled = isAssembled;
        isAssembled = affectedBlocks != null && !affectedBlocks.isEmpty();
        if (affectedBlocks == null || affectedBlocks.isEmpty()) {
            for (BlockPos pos : old) {
                MultiblockCasingCache.removeCasing(serverLevel, pos);
                serverLevel.invalidateCapabilities(pos);
            }
            invalidateCapabilities();
            if (!old.isEmpty()) sync();
            return;
        }
        for (BlockPos pos : old) {
            if (pos.equals(getBlockPos())) continue;
            if (affectedBlocks.contains(pos)) continue;
            MultiblockCasingCache.removeCasing(serverLevel, pos);
            serverLevel.invalidateCapabilities(pos);
        }

        for (BlockPos pos : affectedBlocks) {
            if (pos.equals(getBlockPos())) continue;

            MultiblockCasingCache.setControllerForCasing(serverLevel, pos, getBlockPos());
            serverLevel.invalidateCapabilities(pos);
        }
        invalidateCapabilities();
        blocks = affectedBlocks;
        sync();

        @Nullable BlockPos.MutableBlockPos min = null;
        @Nullable BlockPos.MutableBlockPos max = null;

        for (BlockPos pos : affectedBlocks) {
            if (min == null) min = new BlockPos.MutableBlockPos(pos.getX(), pos.getY(), pos.getZ());
            else {
                if (pos.getX() < min.getX()) min.setX(pos.getX());
                if (pos.getY() < min.getY()) min.setY(pos.getY());
                if (pos.getZ() < min.getZ()) min.setZ(pos.getZ());
            }
            if (max == null) max = new BlockPos.MutableBlockPos(pos.getX(), pos.getY(), pos.getZ());
            else {
                if (pos.getX() > max.getX()) max.setX(pos.getX());
                if (pos.getY() > max.getY()) max.setY(pos.getY());
                if (pos.getZ() > max.getZ()) max.setZ(pos.getZ());
            }
        }
        if (min == null) return;
        max.setX(max.getX() + 1);
        max.setY(max.getY() + 1);
        max.setZ(max.getZ() + 1);

        min.setX(min.getX() - 1);
        min.setY(min.getY() - 1);
        min.setZ(min.getZ() - 1);

        if (wasPreviouslyAssembled) return;

        // redstone particles
        // front + back
        for (int y = min.getY(); y <= max.getY(); ++y) {
            for (int z = min.getZ(); z <= max.getZ(); ++z) {
                // front
                spawnAssembledParticle(min.getX(), y, z, serverLevel);
                // back
                spawnAssembledParticle(max.getX(), y, z, serverLevel);
            }
        }

        // bottom + top
        for (int x = min.getX(); x <= max.getX(); ++x) {
            for (int z = min.getZ(); z <= max.getZ(); ++z) {
                // bottom
                spawnAssembledParticle(x, min.getY(), z, serverLevel);
                // top
                spawnAssembledParticle(x, max.getY(), z, serverLevel);
            }
        }

        // left + right
        for (int y = min.getY(); y <= max.getY(); ++y) {
            for (int x = min.getX(); x <= max.getX(); ++x) {
                // left
                spawnAssembledParticle(x, y, min.getZ(), serverLevel);
                // right
                spawnAssembledParticle(x, y, max.getZ(), serverLevel);
            }
        }
    }

    private void spawnAssembledParticle(int x, int y, int z, ServerLevel level) {
        level.sendParticles(ParticleTypes.END_ROD, x + .5, y + .5, z + .5, 1, 0d, 0d, 0d, 0d);
    }

    public boolean isAssembled() {
        return isAssembled;
    }
}

File: src/main/kotlin/redcrafter07/processed/multiblock/MultiblockBlockEntity.kt
package redcrafter07.processed.multiblock

import net.minecraft.client.multiplayer.ClientLevel
import net.minecraft.core.BlockPos
import net.minecraft.core.BlockPos.MutableBlockPos
import net.minecraft.core.HolderLookup
import net.minecraft.core.particles.ParticleTypes
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.Tag
import net.minecraft.network.chat.Component
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockState
import net.neoforged.neoforge.items.IItemHandler
import redcrafter07.processed.block.machine_abstractions.BlockSide
import redcrafter07.processed.block.machine_abstractions.ProcessedMachine
import redcrafter07.processed.getFacingDirection
import redcrafter07.processed.network.RPCFunctions

abstract class MultiblockBlockEntity(type: BlockEntityType<*>, pos: BlockPos, blockState: BlockState) :
    ProcessedMachine(type, pos, blockState) {
    protected abstract fun validator(): MultiblockValidator

    private var timeUntilNextCheck = 40

    // The parts of this multiblock that were destroyed last tick.
    private val partsDestroyed: ArrayList<BlockPos> = ArrayList()
    var isAssembled: Boolean = false
        private set
    private var blocks: Set<BlockPos>? = null
    protected var specialBlocks: Map<SpecialBlockType, List<BlockPos>> = mapOf()

    protected fun specialBlock(type: SpecialBlockType) =
        specialBlocks[type].run { if (this.isNullOrEmpty()) null else this[0] }

    open fun state(): Component? = null

    override fun loadAdditional(tag: CompoundTag, registries: HolderLookup.Provider) {
        super.loadAdditional(tag, registries)
        isAssembled = tag.getBoolean("isAssembled")

        // blocks are stored relative to the controller as packed longs. This is fine because the MultiBlockCasingCache also assumes the controller is not too far.
        if (tag.contains("blocks", Tag.TAG_LONG_ARRAY.toInt())) {
            val blocks = mutableSetOf<BlockPos>()
            val x = blockPos.x
            val y = blockPos.y
            val z = blockPos.z
            val longs = tag.getLongArray("blocks")
            longs.map(BlockPos::of).map { pos -> BlockPos(pos.x + x, pos.y + y, pos.z + z) }.forEach(blocks::add)
            this.blocks = blocks

            if (tag.contains("specialBlocks", Tag.TAG_COMPOUND.toInt())) {
                val special = tag.getCompound("specialBlocks")
                val specialBlocks = hashMapOf<SpecialBlockType, List<BlockPos>>()
                for (specialBlock in SpecialBlockType.values) {
                    if (special.contains(specialBlock.key, Tag.TAG_INT.toInt())) {
                        val index = special.getInt(specialBlock.key)
                        val unpacked = BlockPos.of(longs[index])
                        specialBlocks[specialBlock] = listOf(BlockPos(unpacked.x + x, unpacked.y + y, unpacked.z + z))
                    } else if (special.contains(specialBlock.key, Tag.TAG_INT_ARRAY.toInt())) {
                        specialBlocks[specialBlock] = listOf(*special.getIntArray(specialBlock.key).map {
                            val unpacked = BlockPos.of(longs[it])
                            BlockPos(unpacked.x + x, unpacked.y + y, unpacked.z + z)
                        }.toTypedArray())
                    }
                }
                this.specialBlocks =
                    mapOf(*specialBlocks.entries.map { (key, value) -> Pair(key, value) }.toTypedArray())
            } else specialBlocks = mapOf()
        } else blocks = null
    }

    override fun saveAdditional(tag: CompoundTag, provider: HolderLookup.Provider) {
        super.saveAdditional(tag, provider)
        tag.putBoolean("isAssembled", isAssembled)
        val blocks = blocks ?: return
        val x = blockPos.x
        val y = blockPos.y
        val z = blockPos.z
        val packedBlocks = blocks.stream().map { pos -> BlockPos.asLong(pos.x - x, pos.y - y, pos.z - z) }.toList()
        tag.putLongArray("blocks", packedBlocks)
        if (specialBlocks.isNotEmpty()) {
            val special = CompoundTag()
            for (entry in specialBlocks.entries) {
                if (entry.value.isEmpty()) continue
                else if (entry.value.size == 1) {
                    val pos = entry.value[0]
                    val long = BlockPos.asLong(pos.x - x, pos.y - y, pos.z - z)
                    val index = packedBlocks.indexOf(long)
                    special.putInt(entry.key.key, index)
                } else {
                    val array = mutableListOf<Int>()
                    for (pos in entry.value) {
                        val long = BlockPos.asLong(pos.x - x, pos.y - y, pos.z - z)
                        val index = packedBlocks.indexOf(long)
                        array.add(index)
                    }
                    special.putIntArray(entry.key.key, array)
                }
            }
            tag.put("specialBlocks", special)
        }
    }

    fun onRemove(level: Level) {
        val blocks = blocks ?: return
        if (level is ServerLevel && !level.isClientSide()) {
            val x = blockPos.x
            val y = blockPos.y
            val z = blockPos.z
            val relPos = blocks.map { pos -> BlockPos.asLong(pos.x - x, pos.y - y, pos.z - z) }.toLongArray()

            for (player in level.players()) RPCFunctions.notifyMultiblockDestroyed(
                player, relPos, blockPos
            )
        }
        for (pos in blocks) {
            MultiBlockBlockCache.removeBlock(level, pos)
            level.invalidateCapabilities(pos)
        }
    }

    open fun tileTickCommon(level: Level, pos: BlockPos, state: BlockState) {
    }

    open fun tileTickServer(level: ServerLevel, pos: BlockPos, state: BlockState) {
    }

    open fun tileTickClient(level: ClientLevel, pos: BlockPos, state: BlockState) {
    }

    final override fun serverTick(level: ServerLevel, pos: BlockPos, state: BlockState) {
    }

    final override fun clientTick(level: ClientLevel, pos: BlockPos, state: BlockState) {
    }

    final override fun commonTick(level: Level, pos: BlockPos, state: BlockState) {
        if (!isAssembled || !isRunning) return
        tileTickCommon(level, pos, state)
        if (level is ServerLevel && !level.isClientSide()) tileTickServer(level, pos, state)
        else if (level is ClientLevel && level.isClientSide()) tileTickClient(level, pos, state)
    }

    override fun tickNoProcessing(level: Level, pos: BlockPos, state: BlockState) {
        if (level is ServerLevel && !level.isClientSide()) {
            if (partsDestroyed.isNotEmpty()) {
                runRecheck()
                partsDestroyed.clear()
                timeUntilNextCheck = 40
                return
            }
            if (isAssembled) return
            if (timeUntilNextCheck > 0) {
                timeUntilNextCheck -= 1
                return
            }
            timeUntilNextCheck = 40

            runRecheck()
        }
    }

    override fun itemCapabilityForSide(side: BlockSide?, state: BlockState): IItemHandler? {
        if (isAssembled) return super.itemCapabilityForSide(side, state)
        return null
    }

    private val levelServer: ServerLevel?
        get() {
            val level = level
            if (level is ServerLevel && !level.isClientSide) return level
            return null
        }

    /**
     * Notifies the multiblock that a block that's a part of this multiblock was destroyed.
     */
    fun partBlockDestroyed(block: BlockPos) {
        partsDestroyed.add(block)
    }

    private fun runRecheck() {
        val serverLevel = this.levelServer ?: return

        val result = validator().getBlocks(serverLevel, blockPos, getFacingDirection(blockState))
        var affectedBlocks = result?.blocks
        if (affectedBlocks != null) {
            for (block in affectedBlocks) {
                val controllerPos: BlockPos? = MultiBlockBlockCache.getController(serverLevel, block)
                if (controllerPos != null && controllerPos != blockPos) {
                    affectedBlocks = null
                    break
                }
            }
        }

        val old = blocks ?: setOf()
        blocks = null
        val wasPreviouslyAssembled = isAssembled
        isAssembled = affectedBlocks != null && true && !affectedBlocks.isEmpty()
        if (!isAssembled || affectedBlocks == null || result == null) {
            for (pos in old) {
                MultiBlockBlockCache.removeBlock(serverLevel, pos)
                serverLevel.invalidateCapabilities(pos)
            }
            invalidateCapabilities()
            if (!old.isEmpty()) {
                sync()
                val x = blockPos.x
                val y = blockPos.y
                val z = blockPos.z
                val blocks =
                    old.stream().map { p -> BlockPos.asLong(p.x - x, p.y - y, p.z - z) }.toList().toMutableList()
                for (block in partsDestroyed) if (!old.contains(block)) blocks.add(
                    BlockPos.asLong(
                        block.x - x, block.y - y, block.z - z
                    )
                )

                val blocksArray = blocks.toLongArray()
                for (player in serverLevel.players()) RPCFunctions.notifyMultiblockDestroyed(
                    player, blocksArray, blockPos
                )
            }
            return
        }
        for (pos in old) {
            if (pos == blockPos) continue
            if (affectedBlocks.contains(pos)) continue
            MultiBlockBlockCache.removeBlock(serverLevel, pos)
            serverLevel.invalidateCapabilities(pos)
        }

        for (pos in affectedBlocks) {
            if (pos == blockPos) continue

            MultiBlockBlockCache.setController(serverLevel, pos, blockPos)
            serverLevel.invalidateCapabilities(pos)
        }
        invalidateCapabilities()
        blocks = affectedBlocks
        specialBlocks = mapOf(*result.importantBlocks.map { (key, value) -> Pair(key, listOf(*value.toTypedArray())) }
            .toTypedArray())

        if (!old.isEmpty()) {
            val x = blockPos.x
            val y = blockPos.y
            val z = blockPos.z

            val blocks = old.stream().filter { p -> !affectedBlocks.contains(p) }
                .map { p -> BlockPos.asLong(p.x - x, p.y - y, p.z - z) }.toList().toMutableList()

            for (block in partsDestroyed) {
                if (!old.contains(block)) blocks.add(BlockPos.asLong(block.x - x, block.y - y, block.z - z))
            }

            if (blocks.isNotEmpty()) {
                val blocksArray = blocks.toLongArray()
                for (player in serverLevel.players()) RPCFunctions.notifyMultiblockDestroyed(
                    player, blocksArray, blockPos
                )
            }
        }
        sync()

        var min: MutableBlockPos? = null
        var max: MutableBlockPos? = null

        for (pos in affectedBlocks) {
            if (min == null) min = MutableBlockPos(pos.x, pos.y, pos.z)
            else {
                if (pos.x < min.x) min.x = pos.x
                if (pos.y < min.y) min.y = pos.y
                if (pos.z < min.z) min.z = pos.z
            }
            if (max == null) max = MutableBlockPos(pos.x, pos.y, pos.z)
            else {
                if (pos.x > max.x) max.x = pos.x
                if (pos.y > max.y) max.y = pos.y
                if (pos.z > max.z) max.z = pos.z
            }
        }
        if (min == null || max == null) return
        max.x = max.x + 1
        max.y = max.y + 1
        max.z = max.z + 1

        min.x = min.x - 1
        min.y = min.y - 1
        min.z = min.z - 1

        if (wasPreviouslyAssembled) return

        // redstone particles
        // front + back
        for (y in min.y..max.y) {
            for (z in min.z..max.z) {
                // front
                spawnAssembledParticle(min.x, y, z, serverLevel)
                // back
                spawnAssembledParticle(max.x, y, z, serverLevel)
            }
        }

        // bottom + top
        for (x in min.x..max.x) {
            for (z in min.z..max.z) {
                // bottom
                spawnAssembledParticle(x, min.y, z, serverLevel)
                // top
                spawnAssembledParticle(x, max.y, z, serverLevel)
            }
        }

        // left + right
        for (y in min.y..max.y) {
            for (x in min.x..max.x) {
                // left
                spawnAssembledParticle(x, y, min.z, serverLevel)
                // right
                spawnAssembledParticle(x, y, max.z, serverLevel)
            }
        }
    }

    private fun spawnAssembledParticle(x: Int, y: Int, z: Int, level: ServerLevel) {
        level.sendParticles(ParticleTypes.END_ROD, x + .5, y + .5, z + .5, 1, 0.0, 0.0, 0.0, 0.0)
    }

    enum class SpecialBlockType(val key: String) {
        None(""), Ignored(""), EnergyInput("energyIn"), ItemInput("itemIn"), ItemOutput("itemOut"), FluidInput("fluidIn"), FluidOutput(
            "fluidOut"
        );

        companion object {
            val values = entries.filter { it.key.isNotEmpty() }.toList()
        }
    }
}